content/blog/2019-09-25-bril-to-llvm-ocaml.md

```diff
@@ -0,0 +1,151 @@
+
+```
+v: int = const 1;
+v: bool = const true;
+print v;
+```
+
+This would actually be one potential reason to implement an SSA transformation of Bril before transforming to LLVM (some LLVM optimizations can usually optimize away the stack so for the sake of code efficiency it wasn't strictly necessary). If Bril was in SSA form, then every variable would only have 1 type and we could unambiguously figure out which print to use.
+
+## Evaluation
+
+### Code correctness
+
+In order to evaluate wether I had succeeded at correctly creating code that transforms Bril into LLVM IR I wrote a battery of tests to test every possible operation and every possible edge case of every operation. The way I verified correctness was by generating the LLVM code for a Bril file, compiling and linking the LLVM code with the `helpers.c` file into a binary, running the binary and capturing the standard output of the execution. Then, I ran the same Bril code through `brili` and captured the standard output of that execution. I then compared the two standard outs to see if they agreed. 
+
+One thing that I worried about was that I never defined the expected output of a program. So in the unlikely event that the Bril interpreter and my code have a similar bug (like a copy paste error when handling `add` and `sub`) I would not notice this issue.
+
+I had some test programs that were written in Bril and some that were written in Typescript and compiled to Bril. The former allowed me to test some weird kinds of combinations of instructions not possible to generate by compiling Typescript. The latter allowed me to more easily write large programs that did more complicated things to stress test the code generator.
+
+Some other evaluation metrics were considered, such as comparing the speed of a Bril program when run through the interpreter to the speed of that Bril program when compiled and run natively. However, it was decided that these results would not be very meaningful other that to confirm that running code natively is *much* faster than running code in an interpreter.
+
+Additionally, measuring the speed at which Bril programs can be transformed to LLVM as a function of their code size would be an interesting extension to this project. We would likely prefer that this transformation take a linear amount of time in relation to the size of the input program.
+
+### Code design
+
+As mentioned above, one aspect where I felt like this project didn't succeed was in creating a good, strongly typed, OCaml representation of the Bril AST. Additionally, the way the LLVM code was generated was by simply generating strings of LLVM instructions for each Bril opcode. One potential way to improve upon this is to make an OCaml module that describes the LLVM AST and then write code to transform the Bril AST into the LLVM AST. Then we could just convert the LLVM AST to a string and output it to a file. 
```

Comment:
Another alternative would be to use the official OCaml LLVM bindings:
https://github.com/llvm-mirror/llvm/tree/master/bindings/ocaml